pulsar-client-api/src/main/java/org/apache/pulsar/client/api/MessageIdAdv.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.api;
+
+import java.util.BitSet;
+
+/**
+ * The {@link MessageId} interface provided for advanced users.
+ * <p>
+ * All built-in MessageId implementations should be able to be cast to MessageIdAdv.
+ * </p>
+ */
+public interface MessageIdAdv extends MessageId {
+
+    /**
+     * Get the ledger ID.
+     *
+     * @return the ledger ID
+     */
+    long getLedgerId();
```

Comment:
This proposal has been discussed and modified into the current version of this PR in https://lists.apache.org/thread/25rzflmkfmvxhf3my0ombnbpv7bvgy32 and received 3 binding +1s and 2 non-binding +1s in https://lists.apache.org/thread/kmjq6lf1f11mf6qb8onhnlr17n27fcv4.

The PIP is here: https://github.com/apache/pulsar/issues/18950

As I've explained in the PIP:

> These details might be not much useful to **application users**, but they are important to developers of Pulsar and its ecosystems. 

> It will be difficult in future to support this API if storage layer will not be bookie or bookie API will be changed.

Currently `MessageId` is still exposed to users, which does not expose these internal details. The new `MessageIdAdv` interface is only used for covenience. Users should write the following code after this PR:

```java
MessageIdAdv msgIdAdv = (MessageIdAdv) msgId;
if (msgIdAdv != null) {
    // Get the ledger id, entry id directly
} else {
    // Handle the case when the storage layer might change in future
}
```

Before this PR, there is much code like:

```java
MessageIdImpl impl = (MessageIdImpl) msgId;
if (impl != null) {
    // Check if msgId is another implementation...
} else {
    // ...
}
```

You can also see these code references in the changes of this PR. If you're going to revert this PIP, please at least start the discussion in the mail list to hear more voices.

Reply:
we should not let users use `ledgerId/entryId`. what if we replace bookkeeper with other some other storage? in that case, such codebase will be broken and exposing internal data and encouraging users to use it not a good design decision. I didn't review this PIP before but such discussion happened multiple times in past and Pulsar community had rejected this proposal in past. As we have not done any release yet, so, it's better to revert this PR soon.

Reply:
> we should not let users use ledgerId/entryId.

It's an ideal assumption. In real world, there are a lot of code references that use ledgerId/entryId from the `MessageId`. They have to hack into the `pulsar-client` module and cast `MessageId` to a specific implementation.

Reply:
> what if we replace bookkeeper with other some other storage? 

When users need to access the detailed fields, they **should have assumed the storage is BookKeeper**. Otherwise, there is no need to get the storage details from the `MessageId`.

Reply:
> When users need to access the detailed fields, they should have assumed the storage is BookKeeper.

That's an absolutely wrong assumption and they must not do it. MessageId is just a reference-id to Pulsar and it's nothing related to BookKeeper. it took a huge effort to get rid of zookeeper dependency and should not do the same with any public API. 

> It's an ideal assumption. In real world, there are a lot of code references that use ledgerId/entryId from the MessageId. 

As I mentioned earlier, MessageId is a Pulsar reference and no one should extract any information from MessageId and depend on it. And this PR, exactly does the opposite and encourages users to extract and depend on it. We don't want to pay price for user dependency on such an invalid contract and can't make any changes in Pulsar later on.

Reply:
> That's an absolutely wrong assumption and they must not do it.

I don't want to determine whether they must not do it. But the **FACT** is that many users have already done it. The fact is that the trivial `MessageId` interface cannot satisfy their demands.

> encourages users to extract and depend on it.

This PR encourages users that **have already needed to extract to depend on it** to use the new interface, instead of these **unstable XXXImpl classes**.

Reply:
As I mentioned multiple times, it's not about what user does with Pulsar's internal information as far as Pulsar doesn't provide such a contract then it provides us the flexibility to make enhancements in Pulsar. Before we open-sourced Pulsar, we made a similar decision in past in the contract and it takes a long time to correct it and I don't want to see it again.

Reply:
The fact is that Pulsar provides a hard-to-use interface `MessageId`. Then users, including ecosystem developers, have to dig into the implementation details. I don't like saying that these developers use Pulsar in a wrong way. I'd like to say they are limited by the poor interface that Pulsar provided. Keeping the limitation is harmful to the Pulsar community and could make ecosystem applications flaky and easy to break.

Reply:
> it takes a long time to correct it and I don't want to see it again.

The point is, now it's open sourced and many more external applications need such ability to access these internal fields. When Pulsar was not open sourced, the use cases were limited. But things are different now.

Reply:
I don't have much time to keep debating on this. And it already spent me too much time to clarify the motivation and find the code examples.

Let's use dev@apache.org to make decisions and hear more voices.

Reply:
I don't think Pulsra has poor interface or has any limitation. MessageId is just a reference from Pulsar and it should have correct serialization and deserialization methods. I don't know any system which encourages or provide a contract to extract messageId and depend on internal.
So, I don't think it's fair to say that Pulsar has limitations and it has poor APIs or interfaces. Abstraction is an important part of API contract and that's what Pulsar is doing. why keeping MessageID abstract from user can break application? and that's the exact reason why user application should not depend on such abstraction and should not try to hack that abstraction.

Reply:
> When Pulsar was not open sourced, the use cases were limited. But things are different now.

It's not about usecases but it's about learning to do the right things.